_posts/events-magic.md

```diff
@@ -0,0 +1,81 @@
+
+We're not discussing C# events, like delegates or Event Sourcing events. We're examining events published by a publisher to which more than one subscriber could subscribe. Hold on, C# events and delegates match the definition, and I'm hoping that no one publishes Event Sourcing domain events for public consumption. Unfortunately, and that's why disambiguation is crucial. Let's try again:
+
+> Events asynchronously published by a publisher to which more than one subscriber could asynchronously subscribe and independently consume.
+
+Asynchronicity and independence rule out C# delegates and, at the same time, don't impose any technology. For example, we don't want to state that those events must be messages on a queue.
+
+## Believes
+
+In chatting with many people, the belief goes along the lines of: It's the communication protocol that causes coupling.
+
+That means that swapping the communication protocol for something that, on paper, promises less or no coupling is the solution to all the coupling-related issues.
+
+Let me use a couple of examples. Two components, A and B, communicate by storing data in a database table. A writes some data, which B then reads to continue the business process.
+
+Oh! They use the same storage, and if A and B were (micro)services, that's bad. They are coupled with the same technology and many other irrelevant (pardon my candor) things. The decision is to remove that coupling by, e.g., replacing the communication protocol with HTTP, gRPC, some obscure JSON-based stuff, messages on a queue, or events on a broker.
+
+The problem is that coupling doesn't go anywhere, and things might worsen. What was a monolith with some coupling could transform into a distributed monolith with coupling in the same places, creating many new problems.
+
+## Where should we focus instead?
+
+Events should be the natural consequence of something else, not the primary goal. We should focus on [autonomous components and service boundaries](https://milestone.topics.it/2023/05/17/back-to-basics-boundaries.html). Finding service boundaries is the first step. That alone enables the defining of autonomous components. At this point, we'll realize that the natural way for autonomous components to communicate, especially across service boundaries, is by using thin events carrying only identifiers and no other data.
+
+If we start from the end by introducing events first when the service boundaries are ill-defined, we're destined to face the dreaded "how do I get the data stored in that other service that I need here now to perform my task" question. There can only be three answers:
+
+- Let's expose a RESTful API to allow querying for the needed data.
+- Let's share the needed data using the events and messages we're already sending or publishing.
+- Maybe our service boundaries are wrong.
+
+It's tough to accept the truth presented by the third answer, and it's much easier to go with one of the other two, with a predominance for the first.
```

Comment:
I'd highlight this one more, I think this thought deserves it, maybe like this: "It's tough to accept the truth that you got your service boundaries wrong. It's much more tempting to conclude that you can just expose an API to allow querying for the needed data or replicate the data using events we're already publishing."

Reply:
```suggestion
It's tough to accept the truth that service boundaries are wrong. It's much more tempting to conclude that we can expose an API to allow querying for the needed data or replicate it across services using events the system is already publishing.
```

Reply:
These days, I wonder if there's also a tendency to make it a GraphQL API instead of/as well as a RESTful one but I'm not too familiar with GraphQL to make that call.